File: example/android/android-sdk/src/main/java/com/webengage/sdk/android/utils/http/AbstractCacheHelper.java
package com.webengage.sdk.android.utils.http;


import android.content.Context;

import com.webengage.sdk.android.utils.WebEngageConstant;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

abstract class AbstractCacheHelper {
    protected Context context;
    protected RequestObject object;
    public static final List<Interceptor> interceptors = new ArrayList<Interceptor>();

    public AbstractCacheHelper(Context context, RequestObject object) {
        this.context = context.getApplicationContext();
        this.object = object;
    }

    public static void addInterceptor(Interceptor interceptor) {
        synchronized (interceptors) {
            interceptors.add(interceptor);
        }
    }

    /**
     * checks if file is present in cache and has its corresponding SQL entry
     *
     * @return true or false
     */
    abstract boolean isFilePresent();

    /**
     * Checks if file age is within the time limit
     *
     * @return true or false
     */
    abstract boolean isFileNotExpired();

    /**
     * Establishes an URL connection with server with appropriate headers(ETag,LastModified etc)
     * to check whether file has been modified or not
     *
     * @return HttpResponse with notmodified field set to true if response code is 304
     */

    abstract Response validateFile();

    /**
     * returns input stream from the cached file
     *
     * @return
     */
    abstract Response readFromFile(Response response);

    /**
     * Establishes URL connection for the given network object
     *
     * @return
     */
    abstract Response downloadFile();


    abstract Response getCachedResponse();


    abstract byte[] saveFile(Response response);


    protected void closeCachedResponse() {
        Response response = getCachedResponse();
        if(response != null) {
            response.closeErrorStream();
            response.closeInputStream();
        }
    }

    protected String getPathToDirectory() {
        File file = new File(context.getApplicationInfo().dataDir,
                WebEngageConstant.DIR_PATH);
        if (!file.exists()) {
            file.mkdirs();

        }
        return file.getPath();
    }

    protected Response applyCachePolicy() {
        Response response = null;
        try {
            switch (object.getCachePolicy()) {
                case CachePolicy.GET_DATA_FROM_NETWORK_FIRST_ELSE_FROM_CACHE:
                    response = downloadFile();
                    if (response.isReadable())
                        return response;
                    else
                        return readFromFile(response);
                case CachePolicy.GET_VALIDATED_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE:
                    if (isFilePresent()) {
                        if (isFileNotExpired())
                            return readFromFile(response);
                        else {
                            response = validateFile();
                            if (!response.isReadable()) {
                                return readFromFile(response);
                            } else {
                                closeCachedResponse();
                                return response;
                            }

                        }
                    } else {
                        return downloadFile();
                    }
                case CachePolicy.GET_DATA_FROM_NETWORK_ONLY_NO_CACHING://
                    return downloadFile();
                case CachePolicy.GET_DATA_FROM_CACHE_ONLY:
                    return readFromFile(response);
                case CachePolicy.GET_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE:
                    if (isFilePresent())
                        return readFromFile(response);
                    else {
                        return downloadFile();
                    }

                case CachePolicy.GET_VALIDATED_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE_RETURNS_NULL_IF_OFFLINE_AND_DATA_EXPIRED:
                    if (isFilePresent()) {
                        if (isFileNotExpired())
                            return readFromFile(response);
                        else {
                            response = validateFile();
                            if (!response.isReadable()) {
                                if (response.getErrorStream() != null || response.getException() != null) {
                                    closeCachedResponse();
                                    return response;
                                } else {
                                    if (!response.modified()) {
                                        return readFromFile(response);
                                    } else {
                                        closeCachedResponse();
                                        return response;
                                    }
                                }
                            } else {
                                closeCachedResponse();
                                return response;
                            }

                        }
                    } else {
                        return downloadFile();
                    }

            }
        } catch (Exception e) {
            return null;
        }
        return null;
    }


}

File: example/android/we-personalization/src/main/java/com/webengage/personalization/data/network/AbstractCacheHelper.kt
package com.webengage.personalization.data.network

import android.content.Context

import java.util.*

abstract class AbstractCacheHelper(
    var context: Context,
    var requestObject: RequestObject
) {

    /**
     * checks if file is present in cache and has its corresponding SQL entry
     *
     * @return true or false
     */
    abstract fun isFilePresent(): Boolean

    /**
     * Checks if file age is within the time limit
     *
     * @return true or false
     */
    abstract fun isFileNotExpired(): Boolean

    /**
     * Establishes an URL connection with server with appropriate headers(ETag,LastModified etc)
     * to check whether file has been modified or not
     *
     * @return HttpResponse with notmodified field set to true if response code is 304
     */
    abstract fun validateFile(): Response?

    /**
     * returns input stream from the cached file
     *
     * @return
     */
    abstract fun readFromFile(response: Response?): Response?

    /**
     * Establishes URL connection for the given network object
     *
     * @return
     */
    abstract fun downloadFile(): Response?
    abstract fun getCachedResponse(): Response?

    abstract fun saveFile(response: Response): ByteArray?

    private fun closeCachedResponse() {
        val response = getCachedResponse()
        if (response != null) {
            response.closeErrorStream()
            response.closeInputStream()
        }
    }

    fun applyCachePolicy(): Response? {
        var response: Response? = null
        try {
            when (requestObject.cachePolicy) {
                CachePolicy.GET_DATA_FROM_NETWORK_FIRST_ELSE_FROM_CACHE -> {
                    response = downloadFile()
                    return if (response!!.isReadable()) response else readFromFile(response)
                }
                CachePolicy.GET_VALIDATED_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE -> {
                    return if (isFilePresent()) {
                        if (isFileNotExpired()) readFromFile(response) else {
                            response = validateFile()
                            if (!response!!.isReadable()) {
                                readFromFile(response)
                            } else {
                                closeCachedResponse()
                                response
                            }
                        }
                    } else {
                        downloadFile()
                    }
                }
                CachePolicy.GET_DATA_FROM_NETWORK_ONLY_NO_CACHING -> {
                    return downloadFile()
                }
                CachePolicy.GET_DATA_FROM_CACHE_ONLY -> {
                    return readFromFile(response)
                }
                CachePolicy.GET_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE -> {
                    return if (isFilePresent()) readFromFile(
                        response
                    ) else {
                        downloadFile()
                    }
                }
                CachePolicy.GET_VALIDATED_DATA_FROM_CACHE_FIRST_ELSE_DOWNLOAD_AND_CACHE_RETURNS_NULL_IF_OFFLINE_AND_DATA_EXPIRED -> {
                    return if (isFilePresent()) {
                        if (isFileNotExpired()) readFromFile(response) else {
                            response = validateFile()
                            if (!response!!.isReadable()) {
                                if (response.errorStream != null || response.exception != null) {
                                    closeCachedResponse()
                                    response
                                } else {
                                    if (!response.modified()) {
                                        readFromFile(response)
                                    } else {
                                        closeCachedResponse()
                                        response
                                    }
                                }
                            } else {
                                closeCachedResponse()
                                response
                            }
                        }
                    } else {
                        downloadFile()
                    }
                }
            }
        } catch (e: Exception) {
            return null
        }
        return null
    }

    companion object {
        @JvmField
        val interceptors: MutableList<Interceptor> = ArrayList()
        fun addInterceptor(interceptor: Interceptor) {
            synchronized(interceptors) { interceptors.add(interceptor) }
        }
    }

}